hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -1844,19 +1846,24 @@ object PruneDeadFields {
             None
         })
       case InsertFields(old, fields, fieldOrder) =>
+        val oldStruct = old.typ.asInstanceOf[TStruct]
         val depStruct = requestedType.asInstanceOf[TStruct]
         val depFields = depStruct.fieldNames.toSet
         val rebuiltChild = rebuildIR(old, env, memo)
         val preservedChildFields = rebuiltChild.typ.asInstanceOf[TStruct].fieldNames.toSet
-        InsertFields(rebuiltChild,
-          fields.flatMap { case (f, fir) =>
-            if (depFields.contains(f))
-              Some(f -> rebuildIR(fir, env, memo))
-            else {
-              log.info(s"Prune: InsertFields: eliminating field '$f'")
-              None
-            }
-          }, fieldOrder.map(fds => fds.filter(f => depFields.contains(f) || preservedChildFields.contains(f))))
+        if (depFields.isEmpty) {
```

Comment:
in general I don't like these kind of branches -- the general case should handle empty structs. I think the core bug is that `if (depFields.contains(f))` should be `if (depFields.contains(f)) || preservedChildFields.contains(f))`, right?

Reply:
Correct, the `if (depFields.isEmpty)` is unnecessary. It was intended to further prune the particular case I was in, which was:

```
InsertFields(Ref(g, TStruct("a": TArray)), Seq(("a", ....Some linear algebra.....))
```

with a requested type of empty. This bug fix means you have to  keep the linear algebra around so that `a` gets replaced by something of the right type, but since the requested type was empty it seemed silly to do the linear algebra, so I added that `if (depFields.isEmpty)` case. I guess we can maybe just assume some higher up pruning will remove this whole `InsertFields`, so there's no need to try and special case the empty case here?

Reply:
actually, this is also incorrect -- InsertFields has the same missingness has `old`, but `MakeStruct` is always defined.

I think the way to fix this is probably with a rewrite that inserts a SelectFields on top of `rebuiltChild`.

Reply:
Ah yeah, that's a better idea, plus it removes the need to do `if (depFields.contains(f)) || preservedChildFields.contains(f))`. Thanks, I'll try this.

Reply:
Yeah, so I think you want to insert a select fields if there are any fields that are in both the `rebuiltChild` and `insertedField`, and NOT in `depFields`, right?

Reply:
Yeah, I think that's right. Currently have it working with always inserting a `SelectFields`,  am modifying to do this check

Reply:
Seems to work now, the check was like you said:

```
val insertOverwritesUnrequestedButPreservedField = fields.exists{ case (fieldName, _) =>
      preservedChildFields.contains(fieldName) && !depFields.contains(fieldName)
}
```